File: ios/QingKeSchedule/Domain/ScheduleRules.swift
import Foundation

enum ScheduleRules {
    static func gregorianCalendar(timeZone: TimeZone = .current) -> Calendar {
        var calendar = Calendar(identifier: .gregorian)
        calendar.locale = Locale(identifier: "en_US_POSIX")
        calendar.timeZone = timeZone
        calendar.firstWeekday = 2
        calendar.minimumDaysInFirstWeek = 1
        return calendar
    }

    static func localDate(from value: String, calendar: Calendar) -> Date? {
        guard value.range(
            of: #"^[0-9]{4}-[0-9]{2}-[0-9]{2}$"#,
            options: .regularExpression
        ) != nil else {
            return nil
        }

        let parts = value.split(separator: "-", omittingEmptySubsequences: false)
        guard
            parts.count == 3,
            parts[0].count == 4,
            parts[1].count == 2,
            parts[2].count == 2,
            let year = Int(parts[0]),
            let month = Int(parts[1]),
            let day = Int(parts[2])
        else {
            return nil
        }

        var components = DateComponents()
        components.calendar = calendar
        components.timeZone = calendar.timeZone
        components.year = year
        components.month = month
        components.day = day

        guard let date = calendar.date(from: components) else { return nil }
        let resolved = calendar.dateComponents([.year, .month, .day], from: date)
        guard resolved.year == year, resolved.month == month, resolved.day == day else { return nil }
        return date
    }

    static func startOfMondayWeek(containing date: Date, calendar: Calendar) -> Date? {
        let startOfDay = calendar.startOfDay(for: date)
        let sundayBasedWeekday = calendar.component(.weekday, from: startOfDay)
        let daysSinceMonday = sundayBasedWeekday == 1 ? 6 : sundayBasedWeekday - 2
        return calendar.date(byAdding: .day, value: -daysSinceMonday, to: startOfDay)
    }

    static func teachingWeek(
        for date: Date,
        semester: SemesterDTO,
        calendar: Calendar
    ) -> Int? {
        guard
            let semesterStart = localDate(from: semester.startDate, calendar: calendar),
            let semesterMonday = startOfMondayWeek(containing: semesterStart, calendar: calendar),
            let dateMonday = startOfMondayWeek(containing: date, calendar: calendar),
            let dayDifference = calendar.dateComponents(
                [.day],
                from: semesterMonday,
                to: dateMonday
            ).day
        else {
            return nil
        }

        return dayDifference / 7 + 1
    }

    static func isTeachingWeekInSemester(_ week: Int, semester: SemesterDTO) -> Bool {
        (1...semester.totalWeeks).contains(week)
    }

    static func date(
        forTeachingWeek week: Int,
        dayOfWeek: Int,
        semester: SemesterDTO,
        calendar: Calendar
    ) -> Date? {
        guard
            (1...7).contains(dayOfWeek),
            let semesterStart = localDate(from: semester.startDate, calendar: calendar),
            let semesterMonday = startOfMondayWeek(containing: semesterStart, calendar: calendar)
        else {
            return nil
        }

        return calendar.date(
            byAdding: .day,
            value: (week - 1) * 7 + dayOfWeek - 1,
            to: semesterMonday
        )
    }

    static func scheduleApplies(_ schedule: CourseScheduleDTO, inWeek week: Int) -> Bool {
        guard week >= schedule.startWeek, week <= schedule.endWeek else { return false }
        switch schedule.repeat {
        case .every:
            return true
        case .odd:
            return !week.isMultiple(of: 2)
        case .even:
            return week.isMultiple(of: 2)
        }
    }

    static func occurrences(forWeek week: Int, courses: [CourseDTO]) -> [CourseOccurrenceDTO] {
        courses.flatMap { course in
            course.schedules.compactMap { schedule in
                guard scheduleApplies(schedule, inWeek: week) else { return nil }
                return CourseOccurrenceDTO(course: course, schedule: schedule)
            }
        }
    }

    static func occurrences(
        for date: Date,
        semester: SemesterDTO,
        courses: [CourseDTO],
        calendar: Calendar
    ) -> [CourseOccurrenceDTO] {
        guard
            let week = teachingWeek(for: date, semester: semester, calendar: calendar),
            isTeachingWeekInSemester(week, semester: semester)
        else {
            return []
        }

        let sundayBasedWeekday = calendar.component(.weekday, from: date)
        let dayOfWeek = sundayBasedWeekday == 1 ? 7 : sundayBasedWeekday - 1
        let matching = occurrences(forWeek: week, courses: courses).filter {
            $0.schedule.dayOfWeek == dayOfWeek
        }

        return matching.enumerated().sorted { left, right in
            if left.element.schedule.startPeriod != right.element.schedule.startPeriod {
                return left.element.schedule.startPeriod < right.element.schedule.startPeriod
            }
            if left.element.schedule.endPeriod != right.element.schedule.endPeriod {
                return left.element.schedule.endPeriod < right.element.schedule.endPeriod
            }

            let nameOrder = left.element.course.name.compare(
                right.element.course.name,
                options: [],
                range: nil,
                locale: Locale(identifier: "zh_CN")
            )
            if nameOrder != .orderedSame {
                return nameOrder == .orderedAscending
            }
            return left.offset < right.offset
        }.map(\.element)
    }

    static func occurrenceStatus(
        _ occurrence: CourseOccurrenceDTO,
        semester: SemesterDTO,
        now: Date,
        calendar: Calendar
    ) -> CourseStatus {
        guard
            let startPeriod = semester.periods.first(where: {
                $0.number == occurrence.schedule.startPeriod
            }),
            let endPeriod = semester.periods.first(where: {
                $0.number == occurrence.schedule.endPeriod
            }),
            let startMinutes = minutes(from: startPeriod.startTime),
            let endMinutes = minutes(from: endPeriod.endTime)
        else {
            return .upcoming
        }

        let components = calendar.dateComponents([.hour, .minute], from: now)
        guard let hour = components.hour, let minute = components.minute else {
            return .upcoming
        }
        let currentMinutes = hour * 60 + minute

        if currentMinutes < startMinutes { return .upcoming }
        if currentMinutes <= endMinutes { return .ongoing }
        return .finished
    }

    static func periodRangesOverlap(
        _ left: CourseScheduleDTO,
        _ right: CourseScheduleDTO
    ) -> Bool {
        left.startPeriod <= right.endPeriod && right.startPeriod <= left.endPeriod
    }

    static func overlappingWeeks(
        _ left: CourseScheduleDTO,
        _ right: CourseScheduleDTO
    ) -> [Int] {
        let firstWeek = max(left.startWeek, right.startWeek)
        let lastWeek = min(left.endWeek, right.endWeek)
        guard firstWeek <= lastWeek else { return [] }

        return (firstWeek...lastWeek).filter {
            scheduleApplies(left, inWeek: $0) && scheduleApplies(right, inWeek: $0)
        }
    }

    static func schedulesConflict(
        _ left: CourseScheduleDTO,
        _ right: CourseScheduleDTO
    ) -> Bool {
        left.dayOfWeek == right.dayOfWeek
            && periodRangesOverlap(left, right)
            && !overlappingWeeks(left, right).isEmpty
    }

    static func conflicts(
        for candidate: CourseDTO,
        against existingCourses: [CourseDTO]
    ) -> [ScheduleConflictDTO] {
        var conflicts: [ScheduleConflictDTO] = []

        for candidateSchedule in candidate.schedules {
            for existingCourse in existingCourses where existingCourse.id != candidate.id {
                for existingSchedule in existingCourse.schedules {
                    guard schedulesConflict(candidateSchedule, existingSchedule) else { continue }
                    conflicts.append(
                        ScheduleConflictDTO(
                            candidateCourse: candidate,
                            candidateSchedule: candidateSchedule,
                            existingCourse: existingCourse,
                            existingSchedule: existingSchedule,
                            weeks: overlappingWeeks(candidateSchedule, existingSchedule)
                        )
                    )
                }
            }
        }

        return conflicts
    }

    static func minutes(from value: String) -> Int? {
        guard value.range(
            of: #"^([01][0-9]|2[0-3]):[0-5][0-9]$"#,
            options: .regularExpression
        ) != nil else {
            return nil
        }

        let parts = value.split(separator: ":", omittingEmptySubsequences: false)
        guard
            parts.count == 2,
            parts[0].count == 2,
            parts[1].count == 2,
            let hour = Int(parts[0]),
            let minute = Int(parts[1]),
            (0...23).contains(hour),
            (0...59).contains(minute)
        else {
            return nil
        }
        return hour * 60 + minute
    }
}

File: Android/app/src/main/java/com/qingke/schedule/domain/ScheduleRules.kt
package com.qingke.schedule.domain

import java.time.DayOfWeek
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.temporal.ChronoUnit
import java.time.temporal.TemporalAdjusters

data class ScheduleConflict(
    val candidateCourse: Course,
    val candidateSchedule: CourseSchedule,
    val existingCourse: Course,
    val existingSchedule: CourseSchedule,
    val weeks: List<Int>,
)

data class OccurrenceKey(
    val courseIndex: Int,
    val scheduleIndex: Int,
)

data class CourseOccurrence(
    val course: Course,
    val schedule: CourseSchedule,
    val key: OccurrenceKey,
)

enum class CourseStatus {
    FINISHED,
    ONGOING,
    UPCOMING,
}

object ScheduleRules {
    fun teachingWeek(date: LocalDate, semester: Semester): Int {
        val semesterMonday = semester.startDateAsLocalDate()
            .with(TemporalAdjusters.previousOrSame(DayOfWeek.MONDAY))
        val dateMonday = date.with(TemporalAdjusters.previousOrSame(DayOfWeek.MONDAY))
        return ChronoUnit.WEEKS.between(semesterMonday, dateMonday).toInt() + 1
    }

    fun isTeachingWeekInSemester(week: Int, semester: Semester): Boolean =
        week in 1..semester.totalWeeks

    fun dateForTeachingWeek(week: Int, dayOfWeek: Int, semester: Semester): LocalDate? {
        if (dayOfWeek !in 1..7) return null
        val semesterMonday = semester.startDateAsLocalDate()
            .with(TemporalAdjusters.previousOrSame(DayOfWeek.MONDAY))
        return semesterMonday.plusDays(((week - 1) * 7L) + dayOfWeek - 1)
    }

    fun scheduleApplies(schedule: CourseSchedule, week: Int): Boolean =
        week in schedule.startWeek..schedule.endWeek && when (schedule.repeatRule) {
            RepeatRule.EVERY -> true
            RepeatRule.ODD -> week % 2 != 0
            RepeatRule.EVEN -> week % 2 == 0
        }

    fun occurrencesForWeek(week: Int, courses: List<Course>): List<CourseOccurrence> = buildList {
        courses.forEachIndexed { courseIndex, course ->
            course.schedules.forEachIndexed { scheduleIndex, schedule ->
                if (scheduleApplies(schedule, week)) {
                    add(CourseOccurrence(course, schedule, OccurrenceKey(courseIndex, scheduleIndex)))
                }
            }
        }
    }

    fun minutes(value: String): Int? = parseLocalTime(value)?.let { it.hour * 60 + it.minute }

    fun occurrenceStatus(
        occurrence: CourseOccurrence,
        semester: Semester,
        now: LocalDateTime,
    ): CourseStatus {
        val start = semester.periods.firstOrNull { it.number == occurrence.schedule.startPeriod }
            ?.let { minutes(it.startTime) }
        val end = semester.periods.firstOrNull { it.number == occurrence.schedule.endPeriod }
            ?.let { minutes(it.endTime) }
        val current = now.hour * 60 + now.minute
        return when {
            start == null || end == null || current < start -> CourseStatus.UPCOMING
            current <= end -> CourseStatus.ONGOING
            else -> CourseStatus.FINISHED
        }
    }

    fun periodRangesOverlap(left: CourseSchedule, right: CourseSchedule): Boolean =
        left.startPeriod <= right.endPeriod && right.startPeriod <= left.endPeriod

    fun overlappingWeeks(left: CourseSchedule, right: CourseSchedule): List<Int> {
        val firstWeek = maxOf(left.startWeek, right.startWeek)
        val lastWeek = minOf(left.endWeek, right.endWeek)
        return if (firstWeek > lastWeek) emptyList() else (firstWeek..lastWeek).filter {
            scheduleApplies(left, it) && scheduleApplies(right, it)
        }
    }

    fun schedulesConflict(left: CourseSchedule, right: CourseSchedule): Boolean =
        left.dayOfWeek == right.dayOfWeek &&
            periodRangesOverlap(left, right) &&
            overlappingWeeks(left, right).isNotEmpty()

    fun conflicts(candidate: Course, existingCourses: List<Course>): List<ScheduleConflict> = buildList {
        candidate.schedules.forEach { candidateSchedule ->
            existingCourses.filter { it.id != candidate.id }.forEach { existingCourse ->
                existingCourse.schedules.forEach { existingSchedule ->
                    val weeks = overlappingWeeks(candidateSchedule, existingSchedule)
                    if (candidateSchedule.dayOfWeek == existingSchedule.dayOfWeek &&
                        periodRangesOverlap(candidateSchedule, existingSchedule) && weeks.isNotEmpty()
                    ) {
                        add(ScheduleConflict(candidate, candidateSchedule, existingCourse, existingSchedule, weeks))
                    }
                }
            }
        }
    }

    fun parseLocalDate(value: String): LocalDate? =
        runCatching {
            if (!LOCAL_DATE.matches(value) || value.startsWith("0000-")) null else LocalDate.parse(value)
        }.getOrNull()

    fun parseLocalTime(value: String): LocalTime? =
        runCatching {
            if (!LOCAL_TIME.matches(value)) null else LocalTime.parse(value)
        }.getOrNull()

    private fun Semester.startDateAsLocalDate(): LocalDate =
        requireNotNull(parseLocalDate(startDate)) { "Semester startDate must be valid before calculating a teaching week." }

    private val LOCAL_DATE = Regex("^[0-9]{4}-[0-9]{2}-[0-9]{2}$")
    private val LOCAL_TIME = Regex("^([01][0-9]|2[0-3]):[0-5][0-9]$")
}
